### PR TITLE
Introduce `zephyr_get()` to improve variable consistency.

### PR DESCRIPTION
**Pull request by tejlmand**
Zephyr build system allows users to specify various build settings.

Unfortunately, the use of this in Zephyr is inconsistent, see #40389.

With the introduction of sysbuild, this problem only grows, as settings which are only sourced from environment cannot be passed to a single image.

Therefore a `zephyr_get()` function is introduced which allows a uniform way of passing variables, in order of precedence:
- sysbuild (when in use)
- CMake cache variable.
- Environment variable.
- local variable.

Another benefit is that sysbuild settings are placed in a sysbuild shadow cache, which allows to keep them separated from the image own cache settings.
Also it remove the need in sysbuild to know which settings to share, thus reducing maintenance and the risk of such list getting out-of-sync with what the application expects to be shared.

### CODE REVIEW

File: cmake/modules/extensions.cmake [hunk 1]
```diff
@@ -2379,27 +2452,26 @@ function(zephyr_check_cache variable)
...
-
-  elseif(app_cmake_lists)
-    set(${variable} ${app_cmake_lists})
+  else()
+    zephyr_get(${variable})
+  endif()
 
-  elseif(${CACHE_VAR_REQUIRED})
+  if(${CACHE_VAR_REQUIRED} AND NOT DEFINED ${variable})
     message(FATAL_ERROR "${variable} is not being defined on the CMake command-line in the environment or by the app.")
```
Comment (nordicjm) [inline cmake/modules/extensions.cmake:2467]:
```suggestion
    message(FATAL_ERROR "${variable} is not being defined on the CMake command-line, in the environment or by the app.")
```

File: cmake/modules/kconfig.cmake [hunk 2]
```diff
@@ -318,21 +348,20 @@ add_custom_target(config-twister DEPENDS ${DOTCONFIG})
...
 import_kconfig(${KCONFIG_NAMESPACE} ${DOTCONFIG})
 
-# Re-introduce the CLI Kconfig symbols that survived
-foreach (name ${cache_variable_names})
-  if("${name}" MATCHES "^${KCONFIG_NAMESPACE}_")
-    if(DEFINED ${name})
-      set(${name} ${${name}} CACHE STRING "")
-    endif()
+# Cache the CLI Kconfig symbols that survived through Kconfig, prefixed with CLI_.
+# Remove those who might have changed compared to earlier runs, if they no longer appears.
```
Comment (nordicjm) [inline cmake/modules/kconfig.cmake:360]:
```suggestion
# Remove those who might have changed compared to earlier runs, if they no longer appear.
```

File: cmake/toolchain/gnuarmemb/generic.cmake [hunk 3]
```diff
@@ -1,15 +1,6 @@
 # SPDX-License-Identifier: Apache-2.0
 
-# Until we completely deprecate it
-if(NOT DEFINED ENV{GNUARMEMB_TOOLCHAIN_PATH})
-  if(DEFINED ENV{GCCARMEMB_TOOLCHAIN_PATH})
-    message(WARNING "GCCARMEMB_TOOLCHAIN_PATH is deprecated, please use GNUARMEMB_TOOLCHAIN_PATH instead")
```
Comment (nordicjm) [inline cmake/toolchain/gnuarmemb/generic.cmake:6]: Assuming this will need an update to the change log to note it has been removed

**Comment by nordicjm** [comment]
Encountered an issue with this:

1. Configured and built hello_world with: `cmake -GNinja -DBOARD=<boarrd> -DAPP_DIR=.. ../../../share/sysbuild/`
2. Re-configured and built with: `cmake -GNinja -DBOARD=<board> -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -Dhellow_world_CONFIG_BOARD_ENABLE_CPUNET=y -Dhellow_world_CONFIG_MCUBOOT_SIGNATURE_KEY_FILE="/tmp/aa/bootloader/mcuboot/root-rsa-3072.pem" -DAPP_DIR=.. ../../../share/sysbuild/`
3. Re-configured and built with: `cmake -GNinja -DBOARD=<board> -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -Dhellow_world_CONFIG_BOARD_ENABLE_CPUNET=y -Dhellow_world_CONFIG_MCUBOOT_SIGNATURE_KEY_FILE="/tmp/aa/bootloader/mcuboot/root-rsa-3072.pem" -Dmcuboot_CONFIG_MBEDTLS_CFG_FILE="mcuboot-FAKE-cfg.h" -DAPP_DIR=.. ../../../share/sysbuild/`

At point 3, build failed with:
```
/tmp/aa/zephyr/samples/hello_world/_AA/mcuboot/zephyr/misc/generated/extra_kconfig_options.conf:4: warning: malformed string literal in assignment to MBEDTLS_CFG_FILE (defined at /tmp/aa/bootloader/mcuboot/boot/zephyr/Kconfig:164, modules/mbedtls/Kconfig:50, modules/mbedtls/Kconfig:50). Assignment ignored.
Parsing /tmp/aa/bootloader/mcuboot/boot/zephyr/Kconfig
Loaded configuration '/tmp/aa/zephyr/boards/arm/bl5340_dvk/bl5340_dvk_cpuapp_defconfig'
Merged configuration '/tmp/aa/bootloader/mcuboot/boot/zephyr/prj.conf'
Merged configuration '/tmp/aa/bootloader/mcuboot/boot/zephyr/boards/bl5340_dvk_cpuapp.conf'
Merged configuration '/tmp/aa/zephyr/samples/hello_world/_AA/mcuboot/zephyr/misc/generated/extra_kconfig_options.conf'

error: Aborting due to Kconfig warnings

CMake Error at /tmp/aa/zephyr/cmake/modules/kconfig.cmake:323 (message):
  command failed with return code: 1
Call Stack (most recent call first):
  /tmp/aa/zephyr/cmake/modules/zephyr_default.cmake:124 (include)
  /tmp/aa/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  /tmp/aa/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:97 (include_boilerplate)
  CMakeLists.txt:44 (find_package)


-- Configuring incomplete, errors occurred!
See also "/tmp/aa/zephyr/samples/hello_world/_AA/mcuboot/CMakeFiles/CMakeOutput.log".
See also "/tmp/aa/zephyr/samples/hello_world/_AA/mcuboot/CMakeFiles/CMakeError.log".
CMake Error at cmake/modules/sysbuild_extensions.cmake:110 (message):
  CMake configure failed for Zephyr project: mcuboot

  Location: /tmp/aa/bootloader/mcuboot/boot/zephyr/
Call Stack (most recent call first):
  bootloader/CMakeLists.txt:7 (ExternalZephyrProject_Add)
```

And I have just noticed that I made a typo with hello_world - no warning was emitted about this whereas previously it would emit a warning about unused variables

**Comment by tejlmand** [comment]
> ```
> /tmp/aa/zephyr/samples/hello_world/_AA/mcuboot/zephyr/misc/generated/extra_kconfig_options.conf:4: warning: malformed string literal in assignment to MBEDTLS_CFG_FILE (defined at /tmp/aa/bootloader/mcuboot/boot/zephyr/Kconfig:164, modules/mbedtls/Kconfig:50, modules/mbedtls/Kconfig:50). Assignment ignored.
> ...
> 
> ```

because when passing strings to Kconfig you must ensure that quotes are kept as quotes, that is they must be escaped like this:
```
-Dmcuboot_CONFIG_MBEDTLS_CFG_FILE=\"mcuboot-FAKE-cfg.h\" 
```
If the quotes are not escaped, the string will be a regular CMake string which doesn't contain quotes.

> And I have just noticed that I made a typo with hello_world - no warning was emitted about this whereas previously it would emit a warning about unused variables

yes, this is the disadvantage of simply passing variables to images in the new way.
We cannot know if any image will be using the variable.
The alternative is the old style, but there we need to maintain a separate list in sysbuild, which isn't scalable.
https://github.com/zephyrproject-rtos/zephyr/blob/dd69eede4353d6f34165e6cc7c1682109bc49c9e/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L52-L56

Anytime some Zephyr itself or a module needs a variable that can be set broadly this list needs to be updated.
And especially for Zephyr modules, we don't really want to let module specific code / variable sneak into Zephyr.

**Comment by tejlmand** [comment]
closed in favor of #49674

File: cmake/modules/extensions.cmake (hunk 1 above)
Comment (tejlmand) [reply cmake/modules/extensions.cmake:2467]: fixed in #49674

**Pull request closed without merging**